### DISCUSSION TITLE
Tutorial on iCub blink or face emotion?

### DISCUSSION THREAD

**Discussion by YANGGANG1971** [robotology/community#429] [Q&A] [answered]
Now, we want to control the physical robot as naturally as possible. Do you have any tutorials on iCub blink and face emotion? Or any website I  can refer to?

**Comment by pattacini** [comment]
Hi @YANGGANG1971 

Is your robot [iCubOsaka01][1]?
Just asking beforehand as there exist different hardware versions underneath the use of facial expressions.

[1]: https://github.com/robotology/icub-tech-support/issues?q=is%3Aissue+author%3AYANGGANG1971

**Comment by YANGGANG1971** [comment]
yes, iCubOsaka01

**Comment by pattacini** [comment, marked as answer]
Here we go:
- Application scripts for running the face expressions are contained in the [main bundle][1] and described in our [official documentation][2]. While reading the doc, bear in mind that your robot `iCubOsaka01` doesn't mount the latest RFE board.
- We have also developed in the past a specific module controlling the blinking behavior. It's called [`iCubBlinker`][3] and is hosted at https://github.com/robotology/funny-things.

[1]: https://github.com/robotology/icub-main/tree/master/app/faceExpressions
[2]: https://robotology.github.io/robotology-documentation/doc/html/group__icub__faceExpressions.html
[3]: http://robotology.github.io/funny-things/module/iCubBlinker.html

**Comment by YANGGANG1971** [comment]
Thanks. We want to control blink and facial expressions using the C++ code.  It seems your suggestion recommends .sh file or command line to control iCub blink. Is there a similar method using C++ as controls the arm motion as follows?
```c++
        Property optArm;//Enable arm device
        optArm.put("device","cartesiancontrollerclient");
        optArm.put("remote","/icub/cartesianController/right_arm");
        optArm.put("local","/cartesian_client/right_arm");
        drvArm.open(optArm)
        bool ok=false;
        drvArm.open(optArm))
        drvArm.view(iarm);//10 joints
        iarm->storeContext(&startup_ctxt_arm);
        iarm->setTrajTime(0.5); //max. time 
        iarm->goToPose(x,o); 
```

**Comment by pattacini** [comment]
Well, the bash scripts (e.g. [`cycle.sh`][1]) do simply rely on standard port-based communication.

Thereby, the following command:
```sh
echo "set all hap"  | yarp rpc /icub/face/emotions/in
```
has clearly its C++ counterpart:
```c++
yarp::os::BufferedPort<yarp::os::Bottle> port;
port.open("/emotions/out");
yarp::os::Network::connect(port.getName(), "/icub/face/emotions/in", "tcp");

yarp:os::Bottle& cmd = port.prepare();
cmd.clear();
cmd.addString("set all hap");
port.writeStrict();
```

[1]: https://github.com/robotology/icub-main/blob/master/app/faceExpressions/scripts/cycle.sh#L8

**Comment by YANGGANG1971** [comment]
Thanks, I will test it on our physical robot.

**Comment by YANGGANG1971** [comment]
Following your suggestion, it seems something wrong while opening icubDemoScripts to call modules followed by the tutorial
 https://github.com/robotology/funny-things#step-by-step-guide. Before that,  I have executed yarpserver --write and iCub_SIM.

![Screenshot from 2020-09-11 11-27-35](https://user-images.githubusercontent.com/55526431/92845840-5d52e300-f422-11ea-8928-67dd9ddcfd5b.png)

**Comment by YANGGANG1971** [comment]
Also, for the physical robot, the error even occurred at while compiling and see the figure blow.
![Screenshot from 2020-09-11 12-20-18](https://user-images.githubusercontent.com/55526431/92851664-cdfcfe00-f428-11ea-8317-ba0004b4f3da.png)

I guess it is due to yarp version or iCub_main version. In fact, I do not want to update the version, because I have many program relying on the current software version. 
My desktop configuration is as follows,
yarp 3.1.0
Ubuntu 16.04
cmake 3.5.1
swig 3.0.8
icub_main v1.12.0

Do you have any idea for addressing this issue, or other way to realize iCub blink?

**Comment by pattacini** [comment]
Hi @YANGGANG1971 

From the error reported by `yarpmanager`

![image](https://user-images.githubusercontent.com/3738070/92878527-79e42080-f40c-11ea-8fd1-b61a5ce56993.png)

it seems a problem related to the availability of `yarprun`. So, you'd better check this out.

Unfortunately, `funny-things` is NOT versioned according to our [Software Versioning Table][1], so you ought to go back in the `funny-things` Git history up to the point in time when your `yarp` and `icub-main` versions have been delivered. This is a bit tricky but it's the fastest workaround, I believe.

[1]: http://wiki.icub.org/wiki/Software_Versioning_Table

**Comment by kt10aan** [comment]
@YANGGANG1971 , the iCubOsaka01 was last updated around September 2018 if I remember correctly. Hopefully that information will narrow down the git history search.

**Comment by YANGGANG1971** [comment]
> Hi @YANGGANG1971
> 
> From the error reported by `yarpmanager`
> 
> ![image](https://user-images.githubusercontent.com/3738070/92878527-79e42080-f40c-11ea-8fd1-b61a5ce56993.png)
> 
> it seems a problem related to the availability of `yarprun`. So, you'd better check this out.
> 
> Unfortunately, `funny-things` is NOT versioned according to our [Software Versioning Table](http://wiki.icub.org/wiki/Software_Versioning_Table), so you ought to go back in the `funny-things` Git history up to the point in time when your `yarp` and `icub-main` versions have been delivered. This is a bit tricky but it's the fastest workaround, I believe.

I am a little bit confused. yarprun aims at killing, monitoring other programs on diff. set of machines. If I only use the iCub sim. on the local machine, do I need to do yarprun --server /SERVERPORT? If indeed, how do I set the server name? p104?

**Comment by pattacini** [comment]
The `yarprun` is a broker that is required to be running on machine `B` when you aim to launch an application from the machine `A` (that is where you rely on `yarpmanager`, for example) on machine `B`. Therefore, `yarprun` doesn't serve only for killing/monitoring remote applications but primarily for starting/stopping them.

When you want to launch an application on the same machine of `yarpmanager` you may replace `yarprun` broker with `localhost`.

From the snapshot you posted above, only `simFaceExpressions` seems correctly configured to run locally using the broker `localhost`, whereas the remainder of the modules are needing the `terminal` port of the `yarprun` server.

So, either you launch `yarprun --server /terminal` on your shell prior to clicking on <kbd>Run All</kbd> button, or simpler edit the application file to replace `terminal` with `localhost`.

**Comment by YANGGANG1971** [comment]
Thanks, it works. But the new problem occurred as follows.
![Screenshot from 2020-09-14 18-47-26](https://user-images.githubusercontent.com/55526431/93071279-ea17ce00-f6ba-11ea-8fc1-996573a49b91.png)
Error: cannot run gaze.lua on locahost: Can't execute command because no such file or directory. 
I am wondering by default, where gaze.lua should be put or where the .xml will first look for? Or how do I address this issue above?

**Comment by pattacini** [comment]
The correct name of the module is indeed `simFaceExpressions`: fixed in https://github.com/robotology/funny-things/commit/927ffa0efbb1f81485e176a16e87a5a67bdd573a. You ought to edit the application file then.

The LUA script `gaze.lua` should get installed correctly in a location that shall be in the path. So, check it out by doing:
```sh
$ which gaze.lua
```

However, LUA does also need to be installed correctly. Follow these instructions: http://www.yarp.it/yarp_swig.html#yarp_swig_lua.

For testing blinking, `gaze.lua` is not necessary though.

**Comment by YANGGANG1971** [comment]
In iCub Sim, now blinking is working. Thanks, but the gaze.lua still cannot work, although I have installed lua. It seems no affecting the blinking function. 
![Screenshot from 2020-09-15 11-47-31](https://user-images.githubusercontent.com/55526431/93159610-84beee00-f749-11ea-8e1a-fe86040a239b.png)
Later, I will test it on the physical iCub.

**Comment by YANGGANG1971** [comment]
The physical iCub is also working. Thanks.

**Comment by pattacini** [comment]
>The physical iCub is also working. Thanks.

Happy it worked 👍🏻 

The script `gaze.lua` has nothing to do with blinking as it merely asks the module `iKinGazeCtrl` to move the gaze toward some given locations, even randomly.

Did you try out one of the LUA examples available in YARP?
This one is a good start: https://github.com/robotology/yarp/blob/master/bindings/lua/examples/example.lua.

**Comment by YANGGANG1971** [comment]
Thx,  I will try.
